Add warning on ProcessGroup and ProcessGroup::Work APIs

agolynski opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46220 Add warning on ProcessGroup and ProcessGroup::Work APIs**

Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

agolynski committed:
warning

[ghstack-poisoned]

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"




[ghstack-poisoned]

mrshenli reviewed (changes requested):
Hey @agolynski, I might not convey this clearly. The goal is that, whenever users call certain APIs on `Work`, we print the warning message. We will need to add things like `TORCH_WARN_ONCE` to impacted APIs. The impacted ones are those that won't be available in the new Future API. At least including the following ones.


https://github.com/pytorch/pytorch/blob/b98e35948f337b40c366cc2568e4921f2cf02c0d/torch/csrc/distributed/c10d/init.cpp#L979-L981

https://github.com/pytorch/pytorch/blob/b98e35948f337b40c366cc2568e4921f2cf02c0d/torch/csrc/distributed/c10d/init.cpp#L987

@wanchaol @gmagogsfm please comment if there are more impacted APIs.

mrshenli commented:
@agolynski 

Raised the concern that APIs like `source_rank` is also used by other features like `torch.distributed.recv`, so adding a warning message will impact those as well. To avoid this, can we do sth like (this can also apply to other APIs):

1) provide both `source_rank` and `_source_rank` APIs. 
2) change our code to use` _source_rank`
3) let `source_rank` print warning

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"




[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -974,15 +974,48 @@ that adds a prefix to each key inserted to the store.
 
   shared_ptr_class_<::c10d::ProcessGroup::Work>(module, "Work")
       .def("is_completed", &::c10d::ProcessGroup::Work::isCompleted)
```
shall we deprecate this as well? This is called `done` in future, see https://github.com/pytorch/pytorch/blob/a3caa719afc3a6fe17c086cde6d5b4d6837ab96c/torch/csrc/jit/python/init.cpp#L1094

agolynski replied:
done

mrshenli replied:
This is risky, as we did expose this API in our API doc. If you wanna change this, please mention it where it is defined in `https://github.com/pytorch/pytorch/blob/master/docs/source/distributed.rst` as well.

agolynski replied:
undone

wanchaol commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -157,6 +160,12 @@ class ProcessGroup {
     return size_;
   }
 
+  // *************************************************************************
```
shall we add this comment in the beginning of `ProcessGroup` class instead?

agolynski replied:
moved

wanchaol reviewed (approved):
looks good to me, minor comments inline. Thanks for doing this!

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"




[ghstack-poisoned]

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"




[ghstack-poisoned]

agolynski commented:
> Hey @agolynski, I might not convey this clearly. The goal is that, whenever users call certain APIs on `Work`, we print the warning message. We will need to add things like `TORCH_WARN_ONCE` to impacted APIs. The impacted ones are those that won't be available in the new Future API. At least including the following ones.
> 
> https://github.com/pytorch/pytorch/blob/b98e35948f337b40c366cc2568e4921f2cf02c0d/torch/csrc/distributed/c10d/init.cpp#L979-L981
> 
> https://github.com/pytorch/pytorch/blob/b98e35948f337b40c366cc2568e4921f2cf02c0d/torch/csrc/distributed/c10d/init.cpp#L987
> 
> @wanchaol @gmagogsfm please comment if there are more impacted APIs.

done

agolynski commented:
> @agolynski
> 
> Raised the concern that APIs like `source_rank` is also used by other features like `torch.distributed.recv`, so adding a warning message will impact those as well. To avoid this, can we do sth like (this can also apply to other APIs):
> 
> 1. provide both `source_rank` and `_source_rank` APIs.
> 2. change our code to use` _source_rank`
> 3. let `source_rank` print warning

done

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -973,16 +973,57 @@ that adds a prefix to each key inserted to the store.
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.isSuccess();
```
shouldn't this be `isCompleted`?

agolynski replied:
fixed

mrshenli reviewed (approved):
Stamp to unblock, but please fix the error I mentioned inline.

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -973,16 +973,57 @@ that adds a prefix to each key inserted to the store.
+      .def(
+          "is_completed",
+          [](::c10d::ProcessGroup::Work& work) -> bool {
+            TORCH_WARN_ONCE("ProcessGroup::Work::is_completed API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
```
Is this what clang-format gives you? The devserver clang-format gives me this:

```c++
            TORCH_WARN_ONCE(
                "ProcessGroup::Work::is_completed API is being "
                "deprecated, please ping "
                "https://github.com/pytorch/pytorch/issues/46291 "
                "if you see this warning");
```

mrshenli replied:
And ditto for other APIs as well.

agolynski replied:
reformatted via vscode clang-format

malfet commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -973,16 +973,57 @@ that adds a prefix to each key inserted to the store.
-      .def("is_success", &::c10d::ProcessGroup::Work::isSuccess)
-      .def("exception", &::c10d::ProcessGroup::Work::exception)
-      .def("source_rank", &::c10d::ProcessGroup::Work::sourceRank)
+      .def(
+          "is_completed",
+          [](::c10d::ProcessGroup::Work& work) -> bool {
+            TORCH_WARN_ONCE("ProcessGroup::Work::is_completed API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.isSuccess();
+          })
+      .def(
+          "is_success",
+          [](::c10d::ProcessGroup::Work& work) -> bool {
+            TORCH_WARN_ONCE("ProcessGroup::Work::is_success API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.isSuccess();
+          })
+      .def(
+          "exception",
+          [](::c10d::ProcessGroup::Work& work) -> std::exception_ptr {
+            TORCH_WARN_ONCE("ProcessGroup::Work::exception API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.exception();
+          })
+      .def(
+          "source_rank",
+          [](::c10d::ProcessGroup::Work& work) -> int {
+            TORCH_WARN_ONCE("ProcessGroup::Work::source_rank API is being "
+                "deprecated, please ping "
+                "https://github.com/pytorch/pytorch/issues/46291 "
+                "if you see this warning");
+            return work.sourceRank();
```
I don't like preprocessor macro, but it seem to be ideal to avoid string literal duplication, i.e. use something like
```
#define _DEPRECATED_DEF(name, method) \
  def( name, [](::c10d::ProcessGroup::Work& work) -> int { \
  TORCH_WARN_ONCE("ProcessGroup::Work::" name " API is being " \
                                        "deprecated, please ping " \
                                        "https://github.com/pytorch/pytorch/issues/46291 " \
                                        "if you see this warning"); \
   return work.method(); \
   })
```
```suggestion
      ._DEPRECATED_DEF("is_completed", isCompleted)
      ._DEPRECATED_DEF("is_success", isSuccess)
      ._DEPRECATED_DEF("exception", exception)
      ._DEPRECATED_DEF("source_rank", sourceRank)
```

mrshenli replied:
Another option is `constexpr` + `fmt::format`

agolynski replied:
added a macro

agolynski replied:
... and switched to fmt::format since clang-tidy doesn't like macros

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"


Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -972,17 +972,45 @@ that adds a prefix to each key inserted to the store.
 #endif
 
+#define PROCESS_GROUP_DEPRECATION_WARNING(api_method)              \
+do { TORCH_WARN_ONCE(                                              \
```
curious, why do we need a `do-while` here?

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"


Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -972,17 +972,44 @@ that adds a prefix to each key inserted to the store.
 
+#define PROCESS_GROUP_DEPRECATION_WARNING(api_method)                \
+  TORCH_WARN_ONCE(#api_method                                        \
+                  "API is being deprecated, please ping "            \
```
Curious, do you need a space before "API"?

agolynski replied:
fixed

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"


Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

[ghstack-poisoned]

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"


Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

[ghstack-poisoned]

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"


Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

[ghstack-poisoned]

agolynski commented:
Clang-tidy is broken now, fix is on the way:
https://github.com/pytorch/pytorch/pull/46315

mrshenli commented:
> Clang-tidy is broken now, fix is on the way: #46315

Isn't the clang-tidy failure caused by this PR?

```
[
  {
    path: 'torch/csrc/distributed/c10d/init.cpp',
    start_line: 975,
    end_line: 975,
    start_column: 9,
    end_column: 9,
    annotation_level: 'failure',
    message: "[cppcoreguidelines-macro-usage] warning: function-like macro 'PROCESS_GROUP_DEPRECATION_WARNING' used; consider a 'constexpr' template function"
  }
]
```

agolynski commented:
> > Clang-tidy is broken now, fix is on the way: #46315
> 
> Isn't the clang-tidy failure caused by this PR?
> 
> ```
> [
>   {
>     path: 'torch/csrc/distributed/c10d/init.cpp',
>     start_line: 975,
>     end_line: 975,
>     start_column: 9,
>     end_column: 9,
>     annotation_level: 'failure',
>     message: "[cppcoreguidelines-macro-usage] warning: function-like macro 'PROCESS_GROUP_DEPRECATION_WARNING' used; consider a 'constexpr' template function"
>   }
> ]
> ```

it should be fixed now (before tidy run)

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"


Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

[ghstack-poisoned]

agolynski committed:
Update on "Add warning on ProcessGroup and ProcessGroup::Work APIs"


Differential Revision: [D24294437](https://our.internmc.facebook.com/intern/diff/D24294437)

[ghstack-poisoned]
